packages/core/src/PartialExecutionUtils/filterDisabledNodes.ts

```diff
@@ -0,0 +1,19 @@
+import { NodeConnectionType } from 'n8n-workflow';
+
+import { DirectedGraph } from './DirectedGraph';
+
+export function filterDisabledNodes(graph: DirectedGraph): DirectedGraph {
```

Comment:
> As mentioned in the comment above, orphaned nodes survive in the filtered graph, which is surprising and makes filter rather misleading, as filtering applies only to connections for disabled nodes **but not to the actual disabled nodes**?

In the test above the orphaned node is actually not disabled. Only the root node is disabled. I can't re-attach an AI modal to another node, which is the reason why it's orphaned.

Now I could remove all utility nodes that are attached to an orphaned node. But the problem there is that the direction of the connections changes. The root node is a child of all utility nodes. 

![image](https://github.com/user-attachments/assets/90595688-a6d9-4f4a-a297-027ad0f1fc4c)

It's not like "give me all your AI children", it would be "give me all your AI parents".

![image](https://github.com/user-attachments/assets/73aa3c5d-b8fa-49dc-a328-1f509edbcaf1)

But the utility nodes can have other nodes connected to themselves again and I don't know if the direction of the connection direction swaps again for some of them, so following them and deleting everything connected to them could be more error prone due to the design choice of making the utility nodes parents instead of children.

Additionally removing them even though they are not disabled would be misleading in my opinion.

So I can add deletion for AI utilities if they are connected to a disabled root node, but again, they are not disabled. For now I'd avoid it to be honest.

Reply:
Thanks for details! Definitely not worth a fundamental refactor - I was mostly curious as the inner workings of AI nodes are all new to me.